hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -1891,8 +1891,8 @@ protected void consumeFromClauseRoot(SqmRoot<?> sqmRoot) {
 		fromClauseIndex.register( sqmRoot, tableGroup );
 		currentQuerySpec().getFromClause().addRoot( tableGroup );
 
-		consumeExplicitJoins( sqmRoot, tableGroup );
 		consumeReusablePaths( sqmRoot, tableGroup );
+		consumeExplicitJoins( sqmRoot, tableGroup );
```

Comment:
What's the reason for changing the switch here and below? I'm not sure if this is still correct when we first render reusable paths and only then the explicit joins.

`from Entity a right join a.association1 a1 where a.association2.name = 'abc'`

would render as 

```
select *
from tbl a
inner join association2 a2 on a.a2 = a2.id
right join association1 a1 on a.a1 = a1.id
```

which might be fine, but I just want to bring this up so we can think about the consequences of this again. If the tests run through, it's probably correct, as I think @jwgmeligmeyling tried to explicitly provoke cases where join ordering matters in the tests that you moved.

Reply:
Does Hibernate have implicit joins and is `a.association2.name` actually allowed? 😉 

In the example of derefencing a property from the first join node there isn't much that can go wrong, because either the dereference introduces a left join (which does not remove rows) or it generates an inner join (which might remove rows, but only ones that would already be removed regardless of join order).

Now it gets a bit interesting when dereferencing / implicit joining a property from a join node that is joined by a right or full outer join (`a1` in this example).

However, in the first place it I'd think that any dereference/implicit join can only occur after the join node it references. Otherwise you simply end up with invalid SQL. So if these joins need to appear after the join node they reference already, I think actually it doesn't make much of a difference between join nodes that are inner/left joined and ones dat are right/full outer joined.

Reply:
By the way the tests did indeed try to provike cases where join ordering matters, so as long as the tests are just moved and the assertions are not altered it should give a fair bit of confidence in the change.

Reply:
without these change a test in `OuterJoinTest` (if I remember well `testJoinOrderWithRightJoinWithIdDereference`) was failing with the exception `Could not locate TableGroup exception`